### PR TITLE
Support separate public input reading, cleanup asserts, update format of input files

### PR DESCRIPTION
**Pull request by makxenov**

### CODE REVIEW

File: include/nil/blueprint/input_reader.hpp [hunk 1]
```diff
@@ -31,76 +31,121 @@
...
+                }
+                if (node.children.size() > 1) {
+                    error << "Too much";
+                    return false;
+                }
+                if (node.children.size() == 1) {
+                    auto corresponding_elem = [](llvm::EllipticCurveKind curve_kind) {
+                        switch (curve_kind) {
+                            case llvm::ELLIPTIC_CURVE_PALLAS:
+                                return json_elem::PALLAS;
```
Comment (akokoshn) [inline include/nil/blueprint/input_reader.hpp:97]: implicit conversion from `json_elem` to `bool`

File: include/nil/blueprint/input_reader.hpp [hunk 2]
```diff
@@ -31,76 +31,121 @@
...
-                    vector2 =
-                        process_empty_field(arg_field_type, is_private);
+                    vector1 = process_empty_field(arg_field_type, is_private);
+                    vector2 = process_empty_field(arg_field_type, is_private);
                 } else {
-                    if (value.size() != 1 || !value.contains("curve")) {
-                        std::cerr << "error in json value:\n" << value << "\n";
-                        UNREACHABLE("value.size() != 1 || !value.contains(\"curve\")");
+                    if (!check_curve(node, curve_type->getCurveKind())) {
+                        return {};
```
Comment (akokoshn) [inline include/nil/blueprint/input_reader.hpp:126]: may be an error message here?

File: include/nil/blueprint/input_reader.hpp [hunk 3]
```diff
@@ -164,50 +209,69 @@ namespace nil {
...
+                    return false;
+                }
+                if (node.children.size() > 1) {
+                    return false;
+                }
+                if (node.children.size() == 1) {
+                    auto corresponding_elem = [](llvm::GaloisFieldKind kind) {
+                        switch (kind) {
+                            case llvm::GALOIS_FIELD_PALLAS_BASE:
+                                return json_elem::PALLAS_BASE;
```
Comment (akokoshn) [inline include/nil/blueprint/input_reader.hpp:228]: implicit conversion from `json_elem` to `bool`

File: include/nil/blueprint/input_reader.hpp [hunk 4]
```diff
@@ -164,50 +209,69 @@ namespace nil {
...
-
                 if (!has_values) {
                     return process_empty_field(arg_field_type, is_private);
                 }
-
-                if (value.size() != 1 || !value.contains("field")){
-                    std::cerr << "error in json value:\n" << value << "\n";
-                    UNREACHABLE("value.size() != 1 || !value.contains(\"field\")");
+                if (!check_field(node, arg_field_type)) {
+                    return {};
```
Comment (akokoshn) [inline include/nil/blueprint/input_reader.hpp:252]: an error message?

File: include/nil/blueprint/input_reader.hpp [hunk 5]
```diff
@@ -487,74 +697,108 @@ namespace nil {
...
+                            empty_value;
+
+                    if (has_values && !input_elem.is_object()) {
+                        error << "Expected JSON object as a part of an input array, got \"" << input_elem << "\"";
+                        return false;
+                    }
+                    const boost::json::object &arg_obj = has_values ? input_elem.as_object() : empty_object;
+                    std::string signature;
+                    signature_parser sp;
+                    if (has_values) {
```
Comment (akokoshn) [inline include/nil/blueprint/input_reader.hpp:754]: Now we can parse signature only if `has_value`, right?

**Review by akokoshn — CHANGES_REQUESTED** [review]

File: include/nil/blueprint/input_reader.hpp [hunk 6]
```diff
@@ -31,76 +31,121 @@
...
 #include <nil/blueprint/logger.hpp>
 
 namespace nil {
     namespace blueprint {
+
         template<typename BlueprintFieldType, typename var, typename Assignment>
         class InputReader {
         public:
-            InputReader(stack_frame<var> &frame, program_memory<var> &memory, Assignment &assignmnt, LayoutResolver &layout_resolver, bool has_values) :
+            InputReader(stack_frame<var> &frame, program_memory<var> &memory, Assignment *assignmnt, LayoutResolver &layout_resolver, bool has_values) :
```
[the remark this answers is not in the document]
Comment (makxenov) [reply include/nil/blueprint/input_reader.hpp]: It was done initially to allow empty assignment for public input only case. Rewritten using `if constexpr`.

File: include/nil/blueprint/input_reader.hpp (hunk 1 above)
Comment (makxenov) [reply include/nil/blueprint/input_reader.hpp:116]: Nope, this is the return from lambda function

File: include/nil/blueprint/input_reader.hpp (hunk 2 above)
Comment (makxenov) [reply include/nil/blueprint/input_reader.hpp:146]: done

File: include/nil/blueprint/input_reader.hpp (hunk 3 above)
Comment (makxenov) [reply include/nil/blueprint/input_reader.hpp:248]: lambda again

File: include/nil/blueprint/input_reader.hpp (hunk 5 above)
Comment (makxenov) [reply include/nil/blueprint/input_reader.hpp:782]: Yes, that's correct, we don't have a JSON file at all if `!has_values`

**Review by akokoshn — APPROVED** [review]

File: include/nil/blueprint/parser.hpp [hunk 7]
```diff
@@ -1817,6 +1840,8 @@ namespace nil {
         private:
             llvm::LLVMContext context;
             const llvm::BasicBlock *predecessor = nullptr;
+            std::unique_ptr<llvm::Module> module;
+            llvm::Module::iterator circuit_function_it;
```
[the remark this answers is not in the document]
Comment (makxenov) [reply include/nil/blueprint/parser.hpp]: done

**Review by aleasims — APPROVED** [review]

**Pull request merged**